src/EventStore.Core/Services/VNode/LeaderInfoProvider.cs

```diff
@@ -0,0 +1,72 @@
+﻿using System.Net;
+using EventStore.Common.Utils;
+using EventStore.Core.Data;
+using EndPoint = System.Net.EndPoint;
+
+
+namespace EventStore.Core.Services.VNode {
+
+	public class LeaderInfoProvider {
+		private readonly VNodeInfo _nodeInfo;
+		private readonly GossipAdvertiseInfo _gossipInfo;
+		private readonly Cluster.MemberInfo _leaderInfo;
+
+		public LeaderInfoProvider(VNodeInfo nodeInfo, GossipAdvertiseInfo gossipInfo, Cluster.MemberInfo leaderInfo) {
+			Ensure.NotNull(nodeInfo, "nodeInfo");
+			Ensure.NotNull(gossipInfo, "gossipInfo");
+			
+			_nodeInfo = nodeInfo;
+			_gossipInfo = gossipInfo;
+			_leaderInfo = leaderInfo;
+		}
+
+		public (EndPoint AdvertisedTcpEndPoint, bool IsTcpEndPointSecure, EndPoint AdvertisedHttpEndPoint)
+			GetLeaderInfoEndPoints(){
+
+			(EndPoint TcpEndPoint, bool IsTcpEndPointSecure, EndPoint HttpEndPoint, string AdvertiseHost, int AdvertiseHttpPort, int AdvertiseTcpPort) endpoints;
+			if (_leaderInfo != null) {
+				endpoints = (
+					TcpEndPoint: _leaderInfo.ExternalTcpEndPoint ?? _leaderInfo.ExternalSecureTcpEndPoint,
+					IsTcpEndPointSecure: _leaderInfo.ExternalSecureTcpEndPoint != null,
+					HttpEndPoint: _leaderInfo.HttpEndPoint,
+					AdvertiseHost: _leaderInfo.AdvertiseHostToClientAs,
+					AdvertiseHttpPort: _leaderInfo.AdvertiseHttpPortToClientAs,
+					AdvertiseTcpPort: _leaderInfo.AdvertiseTcpPortToClientAs);
+			} else {
+				var externalTcp = _nodeInfo.ExternalTcp ?? _nodeInfo.ExternalSecureTcp;
+				if (externalTcp != null && externalTcp.Address.Equals(IPAddress.Any)) {
+					endpoints = (
+						TcpEndPoint: _gossipInfo.ExternalTcp ?? _gossipInfo.ExternalSecureTcp,
+						IsTcpEndPointSecure:_gossipInfo.ExternalSecureTcp != null,
+						HttpEndPoint: _gossipInfo.HttpEndPoint,
+						AdvertiseHost:_gossipInfo.AdvertiseHostToClientAs,
+						AdvertiseHttpPort: _gossipInfo.AdvertiseHttpPortToClientAs,
+						AdvertiseTcpPort: _gossipInfo.AdvertiseTcpPortToClientAs);
+				} else {
```

Comment:
This `else` block should be removed. If there is no leader info then we should always return the endpoints from `gossipInfo`.

Otherwise, the node returns the incorrect address if it has an non `0.0.0.0` IP address and an advertised address configured. For example the configuration:

```
IntIp: 127.0.0.1
ExtIp: 127.0.0.1
ClusterSize: 3
DiscoverViaDns: false
GossipSeed: 127.0.0.1:1111,127.0.0.1:2222 # These nodes must not be running
Insecure: true
EnableExternalTcp: true
ReadOnlyReplica: true
AdvertiseHostToClientAs: 192.168.0.7
```

Should return `192.168.0.7` as the leader endpoint even though the `ExtIp` is not `0.0.0.0`